### DISCUSSION TITLE
Logging to a database in a Spring Boot environment

### DISCUSSION THREAD

**Discussion by ppkarwasz** [apache/logging-log4j2#2807] [Ideas]
Spring Boot supports automatic connection establishment with many databases (see [Data](https://docs.spring.io/spring-boot/reference/data/index.html)).
It might be tempting to reuse those connections to forward logs directly to a database. Alas, a couple of problems appear:

- Database connections appear very late in an application's lifecycle, while the logging subsystem is required basically from the start. Currently Spring:

  1. At startup relies on the native auto-configuration of logging backends. For Log4j Core this would pick up a `log4j2.xml` that **does** not contain any `${spring:...}` lookups (since the Spring configuration is not available yet).
  2. When the `Environment` becomes available, Spring reconfigures Log4j Core, based on its [`logging.*` properties](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#appendix.application-properties.core).
  3. The database client beans unfortunately become available only after the `ApplicationContext` is refreshed.

- Database clients also support logging. Such log events can not be delivered to the database, since they would cause recursive logging.

What do you think is the best approach to log events through Spring Boot's managed database clients?
Is it even advisable to use the same database client for logging and business?

Some pros and cons I collected:

- **Pros**: The JDBC appender consumes a lot of connections (see #1916). I might be an advantage to attach it to Spring Boot's connection pool instead of handling multiple pools.
- **Cons**: Logging might require different configuration parameters. For example we recommend to use a separate JPA Persistence Unit for logging (see [JPA Appender](https://logging.apache.org/log4j/2.x/manual/appenders.html#JPAAppender)).

**Comment by ppkarwasz** [comment]
One of the possible solutions to retrieve a database client from Spring Boot is:

- Write a singleton bean that upon initialization:

  1. Retrieves the database client (can be auto-wired).
  2. Stores it in the logger context via [LoggerContext.putObject](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/LoggerContext.html#putObject(java.lang.String,java.lang.Object)).

- Write a custom `ConfigurationFactory` that:

  1. If the logger context contains a database client, creates a database appender programmatically,
  2. Otherwise, it delegates initialization to another factory.

**Comment by vy** [comment]
I would not implement such a sophisticated configuration mechanism, because:

- There is not really a compelling reason to do so. Shaving off one extra connection pool, IMHO, is not such a big win for a modern application which is composed of dozens of such pools.
- The solution is pretty complicated.
- The solution has other disadvantages that you already cited.
- Keeping logging system apart simplifies the architecture for both the application and the logging system implementation immensely. Given how complicated the things already are, I would avoid adding more to the existing complexity.

**Reply by ppkarwasz** [reply]
> - Keeping logging system apart simplifies the architecture for both the application and the logging system implementation immensely. Given how complicated the things already are, I would avoid adding more to the existing complexity.

Makes sense, although if writing logs to a database is required, I would still use Spring Boot's configuration properties to create a database client. We can provide some copy/paste examples of such an integration in our documentation.

The real question probably is whether logging to a database is required. In a big cloud deployment logging to `stdout` should be enough, but for smaller deployments, logging to a database might mean less moving parts.

**Comment by hydra3230** [comment]
Firstly, I don't think that writing logs into database synchronously and fully is valid requirement and design.
Secondly, assuming that we really need to do so, it should be asyschronously with a independent thread, retrive your log key data and message in your own logic implementation.
Spring Boot framework has not designed for this scenario, log4j2 works and loaded following Spring class load design.

If writing key logs into database for key logic and key business, there's no Spring database client issues, just use MDC or NDC of Log4j to do it.
